lib/outdated.js

```diff
@@ -1,437 +1,244 @@
-    var dep = p[0]
-    var depname = p[1]
-    var dir = dep.path
-    var has = p[2]
-    var want = p[3]
-    var latest = p[4]
-    var type = p[6]
-    if (!opts.global) {
-      dir = path.relative(process.cwd(), dir)
-    }
-    out[depname] = { current: has,
-      wanted: want,
-      latest: latest,
-      location: dir
+    if (opts.color) {
+      outTable[0] = outTable[0].map(heading => styles.underline(heading))
     }
-    if (long) {
-      out[depname].type = type
-      out[depname].homepage = dep.package.homepage
-    }
-  })
-  return JSON.stringify(out, null, 2)
-}
 
-function outdated_ (args, path, tree, parentHas, depth, opts, cb) {
-  if (!tree.package) tree.package = {}
-  if (path && moduleName(tree)) path += ' > ' + tree.package.name
-  if (!path && moduleName(tree)) path = tree.package.name
-  if (depth > opts.depth) {
-    return cb(null, [])
+    const tableOpts = {
+      align: ['l', 'r', 'r', 'r', 'l'],
+      stringLength: s => ansiTrim(s).length
+    }
+    output(table(outTable, tableOpts))
   }
-  var types = {}
-  var pkg = tree.package
 
-  if (!tree.children) tree.children = []
+  process.exitCode = outdated.length ? 1 : 0
+}
 
-  var deps = tree.error ? tree.children : tree.children.filter((child) => !isExtraneous(child))
+async function outdated_ (tree, deps, opts) {
+  const list = []
 
-  deps.forEach(function (dep) {
-    types[moduleName(dep)] = 'dependencies'
-  })
+  async function getOutdatedInfo (name, node) {
+    const spec = npa(name)
 
-  Object.keys(tree.missingDeps || {}).forEach(function (name) {
-    deps.push({
-      package: { name: name },
-      path: tree.path,
-      parent: tree,
-      isMissing: true
-    })
-    types[name] = 'dependencies'
-  })
+    // dep is on disk
+    let path, location, current
+    if (node) {
+      path = node.path
+      location = node.location
+      current = node.package.version
+    }
 
-  // If we explicitly asked for dev deps OR we didn't ask for production deps
-  // AND we asked to save dev-deps OR we didn't ask to save anything that's NOT
-  // dev deps then…
-  // (All the save checking here is because this gets called from npm-update currently
-  // and that requires this logic around dev deps.)
-  // FIXME: Refactor npm update to not be in terms of outdated.
-  var dev = opts.dev || /^dev(elopment)?$/.test(opts.also)
-  var prod = opts.production || /^prod(uction)?$/.test(opts.only)
-  if (
-    (dev || !prod) &&
-    (
-      opts['save-dev'] || (!opts.save && !opts['save-optional'])
+    const { package: pkg } = tree
+    const prod = pkg.dependencies
+    const dev = pkg.devDependencies
+    const peer = pkg.peerDependencies
+    const optional = pkg.optionalDependencies
+
+    const type = (
+      (prod && prod[name] && 'dependencies') ||
+      (dev && dev[name] && 'devDependencies') ||
+      (peer && peer[name] && 'peerDependencies') ||
+      (optional && optional[name] && 'optionalDependencies') ||
+      'dependencies'
```

Comment:
Hmm... also, seems like this check only works if you're not doing `--all`, since a lot of the nodes in `tree.inventory` won't be a dependency of the root node.

If you _are_ doing `--all`, then we have a set of dependent edges listed in the `node.edgesOut` set, and figuring out the "wanted" version gets a bit more involved.

Eg, maybe you have a node at `node_modules/foo`, which is version 1.0.0.  The root depends on `1.0`, and a module at `node_modules/bar` depends on `1.x`.  If `foo@1.1.0` is available, it's outdated-and-wanted by `node_modules/bar`, but outdated-and-unwanted by the root node.